Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -1,186 +1,292 @@
+            matrixSmoothNumbers
+              -- Minimise length of matrix
+              | numberOfConstraints < length mat = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ take (numberOfConstraints + 5 * (k + 1)) smoothNumbers
+              | otherwise                        = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ goSelfInitSieving smoothNumbers otherCoeffs
+              where
+                numberOfConstraints = S.size $ foldr (\col acc -> acc <> I.keysSet col) mempty mat
+                mat = trace ("Log filtering: " ++ show (V.length (V.filter (< h) sievedLogInterval), length newSmoothNumbers)) $ fmap snd smoothNumbers
+          pure matrixSmoothNumbers
 
-    goSolving :: Int -> [(Integer, SignedPrimeIntSet)] -> [(Integer, Integer)]
-    goSolving seed sievingData = firstSquare `seq` secondSquare `seq` (firstSquare, secondSquare) : goSolving (seed + 1) sievingData
+    sievingData = removeRows $ goSieving [] initialDecompositionOfA
+    matrix = translate $ fmap (convertToSet . snd) sievingData
+
+    goSolving :: Int -> Int -> [(Integer, Integer)]
+    goSolving seed counter
+      | counter < 5 = firstSquare `seq` secondSquare `seq` (firstSquare, secondSquare) : goSolving (seed + 1) (counter + 1)
+      | otherwise   = findSquares n $ QuadraticSieveConfig t (m + 100 * (k + 1) * (k + 1)) k h
       where
-        firstSquare = findFirstSquare n (V.fromList (fmap fst sievingData)) solution
-        secondSquare = findSecondSquare n (V.fromList (fmap (snd . second primeSet) sievingData)) solution
-        solution = convertToList $ linearSolve' seed $ translate $ fmap snd sievingData
+        firstSquare = findFirstSquare n (fmap fst squaresData)
+        secondSquare = findSecondSquare n (fmap snd squaresData)
+        squaresData = map (sievingData !!) solution
+        solution = withSomeKnown (convertToList . linearSolve seed) matrix
+
+  pure $ trace ("Size of Matrix: " ++ show (withSomeKnown intVal matrix)) $ goSolving (integerToInt n) 0
+
+generatePrimes :: Integer -> Integer -> Int -> [Prime Integer]
+generatePrimes n midPoint len
+  | len <= 0  = []
+  | otherwise = lowerPrimes ++ higherPrimes
+  where
+    higherPrimes = take (len - length lowerPrimes) $ filter positiveResidue $ generatePrimesForwards $ midPoint + 1
+    -- The length of @lowerPrimes@ may not be @len `div` 2@
+    lowerPrimes = take (len `div` 2) $ filter positiveResidue $ generatePrimesBackwards midPoint
+    positiveResidue p = jacobi n (unPrime p) == One
+
+generatePrimesForwards :: Integer -> [Prime Integer]
+generatePrimesForwards from = [nextPrime (max 3 from)..]
+
+generatePrimesBackwards :: Integer -> [Prime Integer]
+generatePrimesBackwards to
+  -- 2 cannot be a factor of a
+  | to <= 2   = []
+  | otherwise = precPrime to : generatePrimesBackwards (unPrime (precPrime to) - 1)
+
+isPrimeInt :: Prime Integer -> Maybe (Prime Integer)
+isPrimeInt x = fromJust . toPrimeIntegral <$> primeInt
```

Comment:
If `primeInt` is not `Nothing` you can simply return `x`, avoiding extra operations: `primeInt >> pure x`.